File: 2023/08/b.py
scheme = ""
desertMap = {}
pos = []

with open('./b-input.txt') as file:
    i = 0
    for line in file:
        line = line.rstrip()
        
        if i == 0:
            scheme = line
        elif i >= 2:
            key = line[0:line.index(" =")]
            value = (line[line.index("(") + 1:line.index(",")], line[line.index(", ") + 2:line.index(")")])
            desertMap[key] = value

            if key.endswith('A'):
                pos.append(key)

        i += 1

importOld = True
if importOld:
    depth = 1487723409
    pos = ['SCZ', 'LFC', 'HLJ', 'PTZ', 'NQZ', 'SLK']
else:
    depth = 0

while True:

    zCounter = 0
    move = depth % len(scheme)
    side = 0
    if scheme[move] == "R":
        side = 1

    nextSteps = []
    for p in pos:
        nextP = desertMap[p][side]
        nextSteps.append(nextP)
        if nextP.endswith('Z'):
            zCounter += 1

    pos = nextSteps.copy()
    depth += 1

    if zCounter >= 3:
        print(depth, pos, scheme[move], zCounter * "###   ")

    if zCounter == len(pos):
        break

print(scheme)
print(desertMap)
print(depth)
        

File: 2023/08/b.js
const fs = require('fs');
const fileContent = fs.readFileSync('./b-input.txt', 'utf-8');
const lines = fileContent.split(/\r?\n/);

var scheme = ""
var desertMap = {}
var pos = []
var depth = 0

var i = 0;
for (var line of lines) {
    if (i == 0) {
        scheme = line
    } else if (i >= 2) {
        var key = line.substring(0, line.indexOf(" ="))
        var value1 = line.substring(line.indexOf("(") + 1, line.indexOf(","))
        var value2 = line.substring(line.indexOf(",") + 2, line.indexOf(")"))
        desertMap[key] = [value1, value2]

        if (key.endsWith('A')) {
            pos.push(key)
        }
    }

    i++
}

var importOld = true
if (importOld) {
    var depth = 1777115964223
    pos = [ 'SCZ', 'NJC', 'KPK', 'NQQ', 'NQZ', 'GVZ' ] 
}

while (true) {

    var zCounter = 0;
    var move = depth % scheme.length;
    var side = 0
    if (scheme[move] == "R") {
        side = 1
    }

    var nextSteps = []
    for (p of pos) {
        var nextP = desertMap[p][side]
        nextSteps.push(nextP)
        if (nextP.endsWith('Z')) {
            zCounter++
        }
    }

    pos = [...nextSteps]
    depth++

    if (zCounter >= 3) {
        console.log(depth, pos, scheme[move], "###   ".repeat(zCounter))
    }

    if (zCounter == pos.length) {
        break
    }
}

console.log(scheme)
console.log(desertMap)
console.log(depth)